libvast/src/system/indexer_downstream_manager.cpp

```diff
@@ -56,6 +56,33 @@ size_t indexer_downstream_manager::buffered(caf::stream_slot slot) const
   return 0u;
 }
 
+void indexer_downstream_manager::close() {
+  VAST_DEBUG(self(), "closing downstream manager");
+  if (closing)
+    return;
+  // At this point, we unregister all partitions by...
+  for (auto it = partitions.begin(); it != partitions.end();) {
+    VAST_ASSERT(*it != nullptr);
+    if (buffered(**it) == 0u) {
+      // ... either removing them directly if the buffers are empty,
+      // meaning all table slices have been forwarded to the indexers,...
```

Comment:
Is this part an optimization, or necessary for correctness? I was wondering why we need to distinguish between empty/non-empty here and in `unregister_partition()`, but not in `register_partition()` below.

Reply:
I believe both instances are purely defensive, I don't want to rely on `emit_batches_impl()` to clean up partitions that are already done from its point of view (although it currently does).